### DISCUSSION TITLE
Output or view the variable value at the quadrature point

### DISCUSSION THREAD

**Discussion by keassyguang** [idaholab/moose#28912] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi everyone,

I'd like to know if it's possible to output or view a variable value at the quadrature point through Paraview? Just like in ABAQUS, we can choose where we want to view, for example the element, node, integration point.
Many thanks.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Yes you need to use the Positions system to pick the locations to compute at. 

Then use this VPP

https://mooseframework.inl.gov/source/vectorpostprocessors/PositionsFunctorValueSampler.html
A functor can be a variable or a function or a functor material property

**Reply by keassyguang** [reply]
Thanks a lot, I will have try later.

**Comment by keassyguang** [comment]
I might find an issue, if the specified output block is "growing" over the time during the simulation (element activated with CoupledVarThresholdElementSubdomainModifier method), the output quadrature points are as the initial defined block rather than the updated block.

**Reply by GiudGiud** [reply]
to update positions, specify an "execute_on" parameter

**Comment by keassyguang** [comment]
> to update positions, specify an "execute_on" parameter

I specified `execute_on = 'TIMESTEP_END'`, and it works now, thanks a lot

**Reply by keassyguang** [reply]
It seems that the variable values at the quadrature point are correct. However, I have some problems when viewing the results in Paraview. Here, I defined the light intensity to decay exponentially from the top surface to the bottom inherited from Material class. The top boundary is defined using the CoupledVarThresholdElementSubdomainModifier and moves upwards every 4 seconds with a layer thickness of 1. I got three problems:

1. The moving boundary (sideset_top) moves correctly when I set `expression = 'y-ceil(t/4.0001)'` for the AuxVariable layerY, but it doesn't move correctly when I set `expression = 'y-ceil(t/4)'`. I know it is a numerical error problem, and that is why I choose 4.0001 instead of 4. However, for the LightIntensityMaterial.C, I also tried to avoid this numerical error problem. As it is implemented below, shen the time comes to 4s, the height of the top surface (domain_height) should be 1, however in the calculation the domain_height is 2. Any good solution to avoid these numerical error issues in the coding?
2. The light intensity on the top surface seems to be wrong except for the last layer. Since we define `light_intensity = 100*exp(-(domain_height - y)/Dp)`, the light intensity for the top surface should always be 100. In the figure below, the light intensity on the upper surface is half of what it should be, even for the one (Time=4s) equals to 30.3 (60.6 = 100*exp(-(2-1)/2)). How can I show the correct value in paraview?

![image(9)](https://github.com/user-attachments/assets/d6d5d448-331c-4e7b-837d-08f34441d2b6)
**header file**
```
#pragma once

#include "ADMaterial.h"

class LightIntensityMaterial : public ADMaterial
{
public:
  static InputParameters validParams();
  LightIntensityMaterial(const InputParameters & parameters);

  virtual void computeQpProperties() override;

protected:
  const Real _I0;
  const Real _Dp;
  const Real _period;
  const Real _I0_on_duration;
  const Real _layer_thickness;
  const Real _x_start;
  const Real _x_end;

  // Light intensity I(x, t)
  ADMaterialProperty<Real> & _I;                  
};
```

**source file**
```
#include "LightIntensityMaterial.h"

registerMooseObject("elementActivationTestApp", LightIntensityMaterial);

InputParameters
LightIntensityMaterial::validParams()
{
  InputParameters params = ADMaterial::validParams();
  params.addRequiredParam<Real>("I0", "Initial light intensity");
  params.addRequiredParam<Real>("Dp", "Penetration depth");
  params.addRequiredParam<Real>("period", "Period over which domain height increases");
  params.addRequiredParam<Real>("I0_on_duration", "Duration during which I0 is on in each period");
  params.addRequiredParam<Real>("layer_thickness", "Domain height increment every period");
  params.addRequiredParam<Real>("x_start", "x_start position");
  params.addRequiredParam<Real>("x_end", "x_end position");
  params.addClassDescription("Kernel to compute monomer light intensity");

  return params;
}

LightIntensityMaterial::LightIntensityMaterial(const InputParameters & parameters)
  : ADMaterial(parameters),
    _I0(getParam<Real>("I0")),
    _Dp(getParam<Real>("Dp")),
    _period(getParam<Real>("period")),
    _I0_on_duration(getParam<Real>("I0_on_duration")),
    _layer_thickness(getParam<Real>("layer_thickness")),
    _x_start(getParam<Real>("x_start")),
    _x_end(getParam<Real>("x_end")),
    _I(declareADProperty<Real>("I"))
{
}

void LightIntensityMaterial::computeQpProperties()
{
  if (_q_point[_qp](0) >= _x_start && _q_point[_qp](0) <= _x_end)
  {
    Real n_periods = 0;
    if (_t == 0){
      n_periods = 1;
    }
    else{
      // Calculate the number of completed height periods
      n_periods = std::ceil(_t / _period);
    }
    
    // Calculate the current domain height
    Real domain_height = n_periods * _layer_thickness;

    // Calculate depth
    Real depth = domain_height - _q_point[_qp](1); // Assuming y is the vertical coordinate (2D case)

    Real t_mod_period = std::fmod(_t, _period);
    ADReal I0_t = (t_mod_period < _I0_on_duration) ? _I0 : 0.0;
    
    _I[_qp] =  I0_t * std::exp(-depth / _Dp);

    // output for debug
    if (1.9<_t && _t<2.1 && _q_point[_qp](0) >= 2.9 && _q_point[_qp](0) <= 3.1 && _q_point[_qp](1) >= 0.9){
      std::cout << "time = " << _t<< std::endl;
      std::cout << "x = " << _q_point[_qp](0) << std::endl;
      std::cout << "y = " << _q_point[_qp](1) << std::endl;
      std::cout << "domain_height = " << domain_height << std::endl;
      std::cout << "depth = " << depth << std::endl;
      std::cout << "I0_t = " << I0_t << std::endl;
      std::cout << "light intensity = " << _I[_qp] << "\n" << std::endl;
    }
  }
  else 
  {
    _I[_qp] = 0.0;
  }
}
```

**input file**
```
[Problem]
  kernel_coverage_check = false
  boundary_restricted_node_integrity_check = false
  boundary_restricted_elem_integrity_check = false
  type = FEProblem
[]
  
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 30
    ny = 20
    xmax = 6
    ymax = 4
  []
  [active_domain] #the initially active domain
    type = SubdomainBoundingBoxGenerator
    input = 'gen'
    block_id = 1
    bottom_left = '0 0 0'
    top_right = '6 1 0'
  []
  [inactive_domain] #the initially inactive domain
    type = SubdomainBoundingBoxGenerator
    input = 'active_domain'
    block_id = 2
    bottom_left = '0 1 0'
    top_right = '6 4 0'
  []
  [sideset_top]
    type = SideSetsAroundSubdomainGenerator
    input = inactive_domain
    normal = '0 1 0'
    block = 1
    new_boundary = 'sideset_top' # top boundary of active domain
  []
[]
  
[Variables]
  [T]
    initial_condition = 400 #Kelvin
    block = 1
  []
[]

[AuxVariables]
  [layerY]
    [AuxKernel]
    type = ParsedAux
    expression = 'y-ceil(t/4.0001)'
    # expression = 'y-ceil(t/4)'
    use_xyzt = true
    execute_on = 'INITIAL TIMESTEP_BEGIN'
    []
  []
  [I_aux]
    block = 1
    order = FIRST
    family = MONOMIAL
  []
[] 
  
[Kernels]
  [conduction]
    block = 1
    type = ADHeatConduction
    variable = T
  []
[]

[AuxKernels]
  [I_aux]
    type = ADMaterialRealAux
    block = 1
    variable = I_aux
    property = I
    execute_on = 'initial timestep_end'
  []
[]

[MeshModifiers]
  [addLayer] 
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'layerY'
    criterion_type = BELOW
    threshold = 0
    subdomain_id = 1
    moving_boundaries = 'sideset_top'
    moving_boundary_subdomain_pairs = '1 2'
    execute_on = 'TIMESTEP_BEGIN'
  []
[]

[Materials]
  [thermal]
    type = ADHeatConductionMaterial
    block = 1
    specific_heat = 1e-2
    thermal_conductivity = 1.0e-3
  []
  [density]
    type = ADGenericConstantMaterial
    block = 1
    prop_names = 'density'
    prop_values = '1'
  []
  [dummy_mat_inactive]
    type = GenericConstantMaterial
    block = 2
    prop_names = 'dummy_mat'
    prop_values = '0.0'
  []
  [Light_Intensity_Material]
    type = LightIntensityMaterial
    block = 1
    I0 = 100                      
    Dp = 2              
    period = 4                   
    layer_thickness = 1         
    I0_on_duration = 3           
    x_start = 0                
    x_end = 6                             
  []
[]

[BCs]
  # In 2D, bottom = 0, right = 1, top = 2, left = 3 
  [bottom_BC]
    type = ADDirichletBC
    variable = T
    boundary = 0
    value = 400
  []
  [side_top_BC]
    type = ADConvectiveHeatFluxBC
    variable = T
    boundary = sideset_top
    T_infinity = 300.0
    heat_transfer_coefficient = 10
  []

  [right_BC]
    type = ADConvectiveHeatFluxBC
    variable = T
    boundary = 'right'
    T_infinity = 300.0
    heat_transfer_coefficient = 20
  []

  [left_BC]
    type = ADConvectiveHeatFluxBC
    variable = T
    boundary = left
    T_infinity = 300.0
    heat_transfer_coefficient = 20
  []
[]

[Executioner]
  type = Transient
  solve_type = PJFNK
  petsc_options_iname = '-ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter'
  petsc_options_value = '201                hypre    boomeramg      8'
  end_time = 16
  dt = 0.1
  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-12
[]

[Outputs]
  exodus = true
[]

**Reply by GiudGiud** [reply]
Hello

There's a lot here. I think you should boil this down to a very simple input so it s more understandable what you want and how it differs from what you get. 

Keep in mind 
```
  [I_aux]
    block = 1
    order = FIRST
    family = MONOMIAL
  []
```
is not output in the expected manner to Exodus. So what you may be looking at could just be visualization artefact.
There's a thread on the forum somewhere about how it differs.

I recommend you switch this one to a constant monomial which are output as expected, to make sure the visualization is not the problem

Guillaume

**Reply by keassyguang** [reply]
Sorry for the lack of clarity.
I have tried setting the variable to `order=CONSTANT` and `family = MONOMIAL`. The results seem to be correct, but we lose the gradient inside the element.

I have two problems:

1. In my code, I need to use time (_t in the source file and t in the input file) to define the movement of the top surface to simulate the addition of material in additive manufacturing. For example, I've defined an auxiliary variable, layerY, whose value is set to `expression = 'y-ceil(t/4)'` as I want to move this top boundary every 4 seconds. Based on the current results, it seems that the time (t) sometimes has a numerical error, e.g. if the time is 4s, the value of t is slightly greater (or smaller) than 4, something like 4.0001s (or 3.9999s). It seems difficult to get the exact integer time.

2. I have defined a material property, light intensity, which decreases exponentially from top surface to bottom. The expression is `light_intensity = I_0*exp(-(domain_height - y)/Dp)`, `domain_height` is the y-coordinate of the top surface (layY), and `Dp` is a constant. For each period (4 seconds) `I_0=100` for `0s<time<3s` and `I_0=0` for `3s<time<4s`. For the nodes (or points) on the top surface, the light intensity should be 100. However, from the figure above, the light intensity of the top surface equal to 50 (half of 100) for t=1.9s, t=4.1s, and t=8.1s while the result for t=12.1s is correct. Since the nodes on the top surface are shared with the inactive elements when time < 12s, the light intensity values on the top surface are always half of what they should be. I am not sure if the MOOSE averages the values of these shared nodes at the interface of the active and inactive domains.

**Reply by GiudGiud** [reply]
> I have tried setting the variable to order=CONSTANT and family = MONOMIAL. The results seem to be correct, but we lose the gradient inside the element.

Yes but the visualization works as expected and that seems to be your problem with first order monomials. 

> , it seems that the time (t) sometimes has a numerical error, e.g. if the time is 4s, the value of t is slightly greater (or smaller) than 4, something like 4.0001s (or 3.9999s). It seems difficult to get the exact integer time.

That is expected. t is the sum of dts, which are floating point numbers. If you want to hit t = 4s exactly, use 4s as a sync time (in any Outputs object). This will force the time stepper to hit it.
OR 
use multiple time steppers, including a TimeSequenceStepper with 4s as a time to hit. 

> Since the nodes on the top surface are shared with the inactive elements when time < 12s, the light intensity values on the top surface are always half of what they should be. I am not sure if the MOOSE averages the values of these shared nodes at the interface of the active and inactive domains.

Is the light intensity a first order monomial?
Did you check the variable value or are you just reporting what you saw in paraview?

**Reply by keassyguang** [reply]
> That is expected. t is the sum of dts, which are floating point numbers. If you want to hit t = 4s exactly, use 4s as a sync time (in any Outputs object). This will force the time stepper to hit it.
OR
use multiple time steppers, including a TimeSequenceStepper with 4s as a time to hit.

Thank you very much. I will try it later.

> Is the light intensity a first order monomial?

Light intensity (I) is defined as material property, and I use a auxVariable I_aux for output.

> Did you check the variable value or are you just reporting what you saw in paraview?

I didn't check the value, just based on the result shown in the paraview. I think this is the visualization artefact in the paraview as I output the sampled values, they are correct. The picture below shows the result at time=1.9s, all values are correct. The top row is based on FIRST and the second on CONSTANT.

Do we have any ideas to avoid the paraview visualization artefacts except using the CONSTANT MONOMIAL?

![Screenshot from 2024-11-15 09-56-09](https://github.com/user-attachments/assets/294935f6-eb55-4a64-b43f-e78c1a1f67b9)


```
[AuxVariables]
  [I_aux]
    block = 1
    order = FIRST
    # order = CONSTANT
    family = MONOMIAL
  []
[] 

[AuxKernels]
  [I_aux]
    type = ADMaterialRealAux
    block = 1
    variable = I_aux
    property = I
    execute_on = 'initial timestep_end'
  []
[]

[Materials]
  [Light_Intensity_Material]
    type = LightIntensityMaterial
    block = 1
    I0 = 100                      
    Dp = 2              
    period = 4                   
    layer_thickness = 1         
    I0_on_duration = 3           
    x_start = 0                
    x_end = 6                             
  []
[]
```

**Reply by GiudGiud** [reply]
> Do we have any ideas to avoid the paraview visualization artefacts except using the CONSTANT MONOMIAL?

You could use the VTK output to plot it in python, bypassing paraview

**Reply by keassyguang** [reply]
Thank you.
